### ISSUE TITLE
Why the confusion matrix when I test the network only prints the first row?

### ISSUE THREAD

**Issue by giuliohubb** [ultralytics/yolov5#13536]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello!!

I am using YOLOv5 for a small object detection problem. 

When I try to test the model after the training on the test set I use the val.py file with task = test and it works correctly. The only issue is that when it saves the confusion matrix it prints only the first row (Here's two examples, one with normalise = True and one with normalise = False). 

Do you know why this is the case? I tried to look at the code but I've found no issues. I have three classes ['cosmics', 'stars', 'satellites'], cosmics are overrepresented - 63383 instances - followed by stars - 6365 instances - and satellites - 620 instances. The IOU thres is 0.5 and I haven't touched the conf_thres

<img width="837" alt="Image" src="https://github.com/user-attachments/assets/e9955f65-c100-4a14-8f77-a1b83350bc74" />

<img width="843" alt="Image" src="https://github.com/user-attachments/assets/e72c9e32-a91c-4895-9907-84a3508fb33b" />

### Additional

One other thing: 

the image size is 532, in training it has been rounded to 544 because it must be multiple of max stride 32. When I test the network if I put the same image size - 544 - it reaches a mAP = 0.658 while if I put image size = 640 it reaches a mAP = 0.852. Do you have an explanation for this behaviour? In training the image size was 544, reaching an mAP = 0.874. And the confusion matrix in the training is printed correctly (I put it down below). I have seen the documentation many times without finding those answers. Thank you very much for your help. 

<img width="822" alt="Image" src="https://github.com/user-attachments/assets/a7453005-291f-4452-86d9-6455c4ff6981" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @giuliohubb, thank you for reaching out and for your detailed explanation and examples! 😊

This is an automated response to let you know that your issue is being reviewed. An Ultralytics engineer will assist you soon.

In the meantime, here are a few steps you can take to ensure we can better assist you:

- If this is a 🐛 **Bug Report**, please provide a **minimum reproducible example (MRE)**, including a clear description of the steps to reproduce the issue, your environment details (Python version, PyTorch version, etc.), and any relevant code snippets or configurations.
  
- If this is a ❓ **Question**, please share as much detail as possible, including any relevant training logs, dataset details, and the specific commands you are using. You can also verify your setup against our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

## Requirements

Ensure you are using [**Python>=3.8.0**](https://www.python.org/) and have installed all required dependencies from our [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt), including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To check or update your environment, you can follow these steps:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```

## Environments

YOLOv5 supports several verified environments with all dependencies preinstalled, such as:
- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

## Status

Our automated tests ensure that YOLOv5 works correctly across macOS, Windows, and Ubuntu for training, validation, inference, export, and benchmarking. You can check the status here:
<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>

If the issue persists or you discover new details, feel free to update this thread. We appreciate your patience and contributions! 🚀

**Comment by pderrenger** [comment]
@giuliohubb the confusion matrix showing only the first row typically occurs when your model is overwhelmingly predicting one class (likely 'cosmics' due to severe class imbalance). This indicates the model isn't detecting other classes during validation. For small object detection, we recommend:

1. **Class Imbalance Mitigation**: Try `class_weights` in your loss function or oversample minority classes
2. **Input Size Consistency**: Use matching train/val image sizes (544x544 in your case). The mAP increase at 640 suggests your test data benefits from higher resolution, but this should be validated against your training setup

For confusion matrix interpretation, see our [Confusion Matrix guide](https://www.ultralytics.com/glossary/confusion-matrix). To verify predictions, inspect `model.predict()` outputs directly.

**Comment by giuliohubb** [comment]
Thank you for your answer.

'This indicates the model isn't detecting other classes during validation.' 

I don't think this was the case, indeed if you look at the first confusion matrix I sent you can clearly see the colour code in the shells correctly displayed, only the numbers are missing. I think it's an issue in the plots. Also if you look at the second confusion matrix you can see that 1180/6365 stars were misclassified as cosmics, but that means that the rest were correctly classified. 
Additionally during the training the class imbalance was the same, but the confusion matrix has been printed correctly. 
I add here the images from the validation,  related to the prediction, showing that the model isn't detecting only cosmics. 

<img width="713" alt="Image" src="https://github.com/user-attachments/assets/622960c4-6f05-4b47-997f-dda0081dea0f" />

**Comment by pderrenger** [comment]
@giuliohubb thanks for clarifying! You're correct that the normalized confusion matrix (first image) shows color-coded relative class distributions, but numerical values may be too small to display clearly in the first row due to class imbalance. For absolute counts, we recommend using `normalize=False` in the val.py command (as shown in your second matrix). The plotted numbers scale with the normalized values, so low-frequency classes might not show text labels. You can inspect the raw matrix data in the `results/*.json` or `results/*.txt` files for precise counts. 

Regarding image size differences: The mAP variance between 544/640px likely stems from your small objects benefiting from higher resolution during inference. For optimal results, maintain consistent train/val image sizes.
